Fix Ephember cannot connect

roberty99 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Update library to pyephember2 to 0.4.3 which has MQTT fix for API version to use. 
Moved from pyephember as owner is no longer active. 
https://github.com/roberty99/pyephember2

Update Climate.py to remove unused calls and also fix the HotWater to always be device id 4. 

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #125374 
- This PR is related to issue: #125374 
- Link to documentation pull request: 
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] I have followed the [perfect PR recommendations][perfect-pr]
- [X] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [X] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [X] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

roberty99 committed:
PyEphEmber update

roberty99 committed:
["pyephember==0.4.1"]

roberty99 committed:
Manifest to use pyephember 0.4.1 version

roberty99 committed:
Merge branch 'dev' into EPHEmber-Fix

MartinHjelmare reviewed (changes requested):
Please read and fill in the PR checklist correctly according to what you have done. You're missing some required steps there.

roberty99 commented:
The owner of pyephember isn't online for a number of years. Can you advise on what's the best course of action to take ?  My account has access on the pypi to update the package there but not the GitHub for it. > Please read and fill in the PR checklist correctly according to what you have done. You're missing some required steps there.

MartinHjelmare commented:
If you have tried to contact the original author, asking them to share or transfer maintenance to you, without success, I suggest you rename the package and publish it as a new project on GitHub and PyPI.

Give credit to the original author in the renamed project but publish it as a separate project, not a fork, since you're not planning to continue development of the original project.

We require 3rd party dependencies to make the source available in a public online code repository, eg GitHub, and releases published on PyPI should correspond to the releases on the public online code repository, such as GitHub.

roberty99 commented:
Thank you.
I have complete this now as package pyephember2. If this is okay I will
proceed to update the integration to use this package instead.

On Tue 11 Mar 2025, 23:29 Martin Hjelmare, ***@***.***> wrote:

> If you have tried to contact the original author, asking them to share or
> transfer maintenance to you, without success, I suggest you rename the
> package and publish it as a new project on GitHub and PyPI.
>
> Give credit to the original author in the renamed project but publish it
> as a separate project, not a fork, since you're not planning to continue
> development of the original project.
>
> We require 3rd party dependencies to make the source available in a public
> online code repository, eg GitHub, and releases published on PyPI should
> correspond to the releases on the public online code repository, such as
> GitHub.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/core/pull/140367#issuecomment-2715950418>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ALVJ7EOX5I33L7562VO3XNT2T5WUHAVCNFSM6AAAAABYY2SFJCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMJVHE2TANBRHA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
> [image: MartinHjelmare]*MartinHjelmare* left a comment
> (home-assistant/core#140367)
> <https://github.com/home-assistant/core/pull/140367#issuecomment-2715950418>
>
> If you have tried to contact the original author, asking them to share or
> transfer maintenance to you, without success, I suggest you rename the
> package and publish it as a new project on GitHub and PyPI.
>
> Give credit to the original author in the renamed project but publish it
> as a separate project, not a fork, since you're not planning to continue
> development of the original project.
>
> We require 3rd party dependencies to make the source available in a public
> online code repository, eg GitHub, and releases published on PyPI should
> correspond to the releases on the public online code repository, such as
> GitHub.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/core/pull/140367#issuecomment-2715950418>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ALVJ7EOX5I33L7562VO3XNT2T5WUHAVCNFSM6AAAAABYY2SFJCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMJVHE2TANBRHA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

MartinHjelmare commented:
Yes, that's ok.

roberty99 committed:
Update requirements_all.txt

roberty99 committed:
updated everything to pyephember2

roberty99 committed:
Merge branch 'dev' into EPHEmber-Fix

roberty99 committed:
Update Clime Platform Schema 

Update Clime Platform Schema `homeassistant.components.climate.PLATFORM_SCHEMA` should be imported as `CLIMATE_PLATFORM_SCHEMA`RuffICN001

roberty99 reviewed (commented):
PLATFORM_SCHEMA as CLIMATE_PLATFORM_SCHEMA on line 22

MartinHjelmare commented on homeassistant/components/ephember/climate.py:
```diff
@@ -33,23 +32,23 @@
+SCAN_INTERVAL = timedelta(seconds=10)
 
-OPERATION_LIST = [HVACMode.HEAT_COOL, HVACMode.HEAT, HVACMode.OFF]
+OPERATION_LIST = [HVACMode.AUTO, HVACMode.HEAT, HVACMode.OFF]
```
Why do we change from `HEAT_COOL` to `AUTO`? The auto mode has specific meaning in Home Assistant.

https://github.com/home-assistant/core/blob/2f1ff5ab95b060d678c1ec0f027460f27c332a02/homeassistant/components/climate/const.py#L21-L23

roberty99 replied:
In this case the thermostat will use the Schedule it has built in.

roberty99 replied:
heat_Cool is never an option for Ember heating systems they are gas boilers.

MartinHjelmare replied:
Is it correct that the user can't adjust the temperature when in the AUTO device mode?

Zaduvalo replied:
eTrv device and some thermostats can change temperature only in Manual/ON mode.
I think AUTO is more accurate mode. In AUTO mode devices use schedule and target temp.
Heat/Cool never exists in EPH devices.

MartinHjelmare replied:
We do not allow using auto mode if temperature can be changed by the user.

MartinHjelmare commented on homeassistant/components/ephember/climate.py:
```diff
@@ -90,7 +90,8 @@ def __init__(self, ember, zone):
         self._zone = zone
-        self._hot_water = zone_is_hot_water(zone)
+        self._hot_water = zone["deviceType"] == 4
+        """4 is a specific device type for immersions returned by EPH. Hot Water temp cannot be changed"""
```
Please split long strings around 88 characters per line.

MartinHjelmare commented on homeassistant/components/ephember/climate.py:
```diff
@@ -33,23 +32,23 @@
+OPERATION_LIST = [HVACMode.AUTO, HVACMode.HEAT, HVACMode.OFF]
 
-PLATFORM_SCHEMA = CLIMATE_PLATFORM_SCHEMA.extend(
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
```
```suggestion
PLATFORM_SCHEMA = CLIMATE_PLATFORM_SCHEMA.extend(
```

roberty99 commented:
They can adjust the temperature when in Auto mode.
Auto just set's the EPHEMber to use it's schedule.
Temp adjust is for Both Auth and Heat modes.

Temperature changing is not allowed on the DeviceType = 4 which is a hot
water heating cycle.
In those cases it should take the temperature that is set on the device and
not allow it to be changed.

Kind Regards,
Robert

On Wed, 12 Mar 2025 at 13:15, Martin Hjelmare ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In homeassistant/components/ephember/climate.py
> <https://github.com/home-assistant/core/pull/140367#discussion_r1991478964>
> :
>
> >
> -OPERATION_LIST = [HVACMode.HEAT_COOL, HVACMode.HEAT, HVACMode.OFF]
> +OPERATION_LIST = [HVACMode.AUTO, HVACMode.HEAT, HVACMode.OFF]
>
> Is it correct that the user can't adjust the temperature when in the AUTO
> device mode?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/core/pull/140367#discussion_r1991478964>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ALVJ7EJMYC64VXOIQIJ2OQ32UAXQJAVCNFSM6AAAAABYY2SFJCVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDMNZYGQ2DKMBZG4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

MartinHjelmare commented:
We can't use `HVACMode.AUTO` in Home Assistant if the user is able to change the temperature in that mode.

roberty99 commented:
![Screenshot_2025-03-12-14-05-26-80_a43fa3331b0de4f6ae70203555620ae1.jpg](https://github.com/user-attachments/assets/3e0ce6a1-4744-4bc0-a742-a137bf02f41d)

The thermostat allows target temp change on Auto. 
Auto dictates the timer schedule to use as opposed to Heat which is always on.

roberty99 commented:
![Screenshot_2025-03-12-14-06-59-92_a43fa3331b0de4f6ae70203555620ae1.jpg](https://github.com/user-attachments/assets/d59bf07a-56e4-46bd-9294-ce4808594486)

roberty99 commented:
I'm not a developer and am out of my depth committing these changes. 
I'm only trying to help as this integration has been broken for 2 + years for the community and I've managed to get a working custom component functioning again with these small changes. 

If we can't fix or support this can we just cancel the official EPHEmber support on HA and we move this to HACS instead ?

MartinHjelmare commented:
Is there a technical problem with keeping the device auto mode mapped to the `HVAC.HEAT_COOL` in the entity as it is today and has been for the last five years? If not, I'd suggest keeping it as such for now.

On another note: We have deprecated `ClimateEntityFeature.AUX_HEAT`.

https://developers.home-assistant.io/blog/2024/03/10/climate-aux-heater-deprecated

The aux heat on/off should be moved to a switch entity if separate control of that feature while the device is on is needed. If it's enough to be able to turn on/off the device the climate entity can support `ClimateEntityFeature.TURN_OFF` and `ClimateEntityFeature.TURN_ON` which can be set to allow the user to turn on/off the entity. When the entity is turned off, no other controls are possible, than to turn it on again.

Removing the aux heat feature can be done in a separate PR.

roberty99 commented:
Heat/Cool: The device is set to heat/cool to a target temperature range.
I'll change to this method now and retest everything.

roberty99 commented:
Using HVACMode.Heat_Cool you need to set High and Low temp range and the thermostat only accepts one target temp to reach. The UI is poor. HVACMode.AUTO is preferrably as we are setting the Thermostat to be Schedule controlled as per the documentation. I did not see why we cannot use AUTO with a target temperature.

roberty99 committed:
PLATFORM_SCHEMA = CLIMATE_PLATFORM_SCHEMA.extend

Zaduvalo commented:
> I have complete this now as package pyephember2. If this is okay I will proceed to update the integration to use this package instead.

Do you think there is no chance to complete the version update in the original ttroy50 repo?
https://github.com/ttroy50/pyephember/pull/19
Just prepare an updated version with new device support and small fixes

roberty99 commented:
Cancelling my PR as this is being submitted here with even more support and fixes: https://github.com/home-assistant/core/pull/140459

roberty99 commented:
Cancelling my PR as this is being submitted here with even more support and fixes: https://github.com/home-assistant/core/pull/140459
